### PR TITLE
Ensure correct types for `QueryBuilder().dict()` with multiple projections

### LINKED ISSUE #3679: Projecting more than just `*` with `QueryBuilder.iterdict` results in `BackendEntity` instances being returned.
Reported by @zooks97

Projecting only `*` when using `dict` works fine:
```
In [11]: QueryBuilder().append(Node, project=['*']).limit(2).dict()
Out[11]: 
[{'node_1': {'*': <Int: uuid: 1984c02b-75c2-43ae-a065-e0a10124977b (pk: 13641) value: 2>}},
 {'node_1': {'*': <FolderData: uuid: edf5d9bf-4dcf-49fb-866f-a38fde52f488 (pk: 112)>}}]
```
However, when adding another projection, the entities are `BackendNode` instances instead of `Node` instances:
```
In [12]: QueryBuilder().append(Node, project=['*', 'id']).limit(2).dict()
Out[12]: 
[{'node_1': {'*': <aiida.orm.implementation.django.nodes.DjangoNode at 0x7ff7f15ffe48>,
   'id': 13641}},
 {'node_1': {'*': <aiida.orm.implementation.django.nodes.DjangoNode at 0x7ff7f15fff28>,
   'id': 112}}]
```
Switching to the `all` resolver works as expected:
```
In [13]: QueryBuilder().append(Node, project=['*', 'id']).limit(2).all()
Out[13]: 
[[<Int: uuid: 1984c02b-75c2-43ae-a065-e0a10124977b (pk: 13641) value: 2>,
  13641],
 [<FolderData: uuid: edf5d9bf-4dcf-49fb-866f-a38fde52f488 (pk: 112)>, 112]]
```

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #3679 

The returned results by the backend implementation of `QueryBuilder` are
passed through `get_aiida_entity_res` to convert backend instances to
front end class instances. However, in the `iterdict` accumulator this
was not accounting for mappings and sequences. This would cause problems
when multiple projections were used, such as `project=['*', 'id']`. The
'*' notation will cause the entire node instance to be loaded, however
due to the fact that it was inside a dictionary in the response it would
not be converted.

### CODE REVIEW

File: aiida/orm/querybuilder.py [hunk 1]
```diff
@@ -1994,15 +1995,20 @@ def get_query(self):
                 self._hash = queryhelp_hash
         return query
 
-    @staticmethod
-    def get_aiida_entity_res(value):
+    def get_aiida_entity_res(self, value):
```
Comment (ltalirz) [inline aiida/orm/querybuilder.py]: could be a `@classmethod`, right?

File: aiida/orm/querybuilder.py [hunk 2]
```diff
@@ -1994,15 +1995,20 @@ def get_query(self):
...
         """Convert a projected query result to front end class if it is an instance of a `BackendEntity`.
 
         Values that are not an `BackendEntity` instance will be returned unaltered
 
         :param value: a projected query result to convert
         :return: the converted value
         """
+        if isinstance(value, collections.abc.Mapping):
+            return {key: self.get_aiida_entity_res(sub_value) for key, sub_value in value.items()}
```
Comment (ltalirz) [inline aiida/orm/querybuilder.py]: I wonder, should this not rather be fixed here:
https://github.com/aiidateam/aiida-core/blob/81d736c3c2f98c3f5ac02a3e6b21224e2e5bedf0/aiida/orm/convert.py#L26-L28

Also, it looks very similar, so I wonder why this is not working...
Is just the list/tuple stuff missing from there?

Comment (sphuber) [reply aiida/orm/querybuilder.py]: You are right, it should be handled there but the problem is that there a `TypeError` can be raised if some of the elements are not node types, which is then caught by `get_aiida_entity_res` and it then returns the original value which is the original dictionary with unconverted values. I will fix it correctly

File: aiida/orm/querybuilder.py (hunk 1 above)
Comment (sphuber) [reply aiida/orm/querybuilder.py]: could be, but I think staticmethod makes more sense. Now that I fixed the problem in the correct place (thanks for catching this :+1:) I think we can keep this as was

**Comment by sphuber** [comment]
> seems pylint tripped over again...
> can we actually fix this warning by using the proper capitalization?

I know, but I cannot reproduce it locally or in other PRs. I have been trying to fix it and already merged two more PRs in a desperate attempt to tackle this. It's infuriating. In any case it is not a problem iwth capitalization but just the fact that it is 32 characters instead of the "max allowed" 30.

**Comment by ltalirz** [comment]
What if we increase the max allowed to 32?

**Comment by sphuber** [comment]
The last commit I did may have done it: https://github.com/aiidateam/aiida-core/runs/358302396
But there is no certainty

**Comment by sphuber** [comment]
> What if we increase the max allowed to 32?

As a last defense I just snuck this in this [last commit](https://github.com/aiidateam/aiida-core/commit/c3c9aaf4d60b58ee633b68bbffdcf40cc1f29ff2). Let's hope to all that is holy that that will do it

**Pull request merged**